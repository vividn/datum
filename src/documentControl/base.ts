import { EitherDocument, EitherPayload } from "./DatumDocument";
import { OutputArgs } from "../input/outputArgs";

export class DocExistsError extends Error {
  data: { existingDocument: EitherDocument; failedPayload: EitherPayload };
  constructor(failedPayload: EitherPayload, existingDocument: EitherDocument) {
    super(
      `conflict: another doc with id "${existingDocument._id}" already exists`,
    );
    this.data = { existingDocument, failedPayload };
    this.name = this.constructor.name;
    Object.setPrototypeOf(this, DocExistsError.prototype);
    Error.captureStackTrace(this, this.constructor);
  }
}

export type BaseDocControlArgs = {
  db: PouchDB.Database<any>;
  outputArgs?: OutputArgs;
};
