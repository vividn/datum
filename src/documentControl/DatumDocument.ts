import {
  isIsoDateOrTime,
  isoDate,
  isoDatetime,
  isoDuration,
} from "../time/timeUtils";
import { WithRequired } from "../utils/utilityTypes";

export type GenericData<T = unknown> = T & {
  [key: string]: any;
};

export type DatumData<T = unknown> = GenericData<T> & {
  occurTime?: isoDatetime | isoDate;
  occurUtcOffset?: number;
  dur?: "start" | "end" | isoDuration;
};

export type OccurredData<T = unknown> = WithRequired<
  DatumData<T>,
  "occurTime" | "occurUtcOffset"
>;

export function isOccurredData(
  data: DatumData | OccurredData
): data is OccurredData {
  const occurTime = data.occurTime;
  return occurTime !== undefined && isIsoDateOrTime(occurTime);
}

export type DatumMetadata = {
  utcOffset?: number;
  createTime?: isoDatetime;
  modifyTime?: isoDatetime; //TODO: turn into an array of times
  idStructure?: string;
  random?: number;
  humanId?: string;
  // [key: string]: any;
};

export type DatumPayload<T = unknown> = {
  _id?: string;
  _rev?: string;
  data: DatumData<T>;
  meta: DatumMetadata;
};

export type DatumDocument<T = unknown> = DatumPayload<T> & {
  _id: string;
  _rev: string;
};

export type DataOnlyPayload<T = unknown> = {
  _id?: string;
  _rev?: string;
} & DatumData<T>;

export type DataOnlyDocument<T = unknown> = DataOnlyPayload<T> & {
  _id: string;
  _rev: string;
};

export type EitherPayload<T = unknown> = DatumPayload<T> | DataOnlyPayload<T>;
export type EitherDocument<T = unknown> =
  | DatumDocument<T>
  | DataOnlyDocument<T>;

export function isDatumDocument(
  doc: DatumDocument | DataOnlyDocument
): doc is DatumDocument {
  return (
    (doc as DatumDocument).data !== undefined &&
    (doc as DatumDocument).meta !== undefined
  );
}

export function isDatumPayload(
  payload: DatumPayload | DataOnlyPayload
): payload is DatumPayload {
  return (
    (payload as DatumPayload).data !== undefined &&
    (payload as DatumPayload).meta !== undefined
  );
}
