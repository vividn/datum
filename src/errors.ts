export class MyError extends Error {
  constructor(m?: unknown) {
    if (typeof m === "string") {
      super(m);
    } else {
      super();
    }
    this.name = this.constructor.name;
    Object.setPrototypeOf(this, MyError.prototype);
  }
}

export class MissingRequiredKeyError extends MyError {
  constructor(key: string) {
    super(`No data given for the required key: ${key}`);
    Object.setPrototypeOf(this, MissingRequiredKeyError.prototype);
  }
}

export class ExtraDataError extends MyError {
  constructor(keys: string[]) {
    super(
      `the following data do not have keys: ${keys.join(
        ", ",
      )}. Assign keys with equals signs, use required/optional keys, specify a key to use as --remainder, or use --lenient`,
    );
    Object.setPrototypeOf(this, ExtraDataError.prototype);
  }
}
export class BaseDataError extends MyError {
  constructor(m: unknown) {
    super(m);
    Object.setPrototypeOf(this, BaseDataError.prototype);
  }
}

export class IdError extends MyError {
  constructor(m: unknown) {
    super(m);
    Object.setPrototypeOf(this, IdError.prototype);
  }
}

export class BadTimeError extends MyError {
  #key = "";
  #value = "";

  constructor(badTime?: string) {
    super();
    this.#value = badTime ?? "";
    Object.setPrototypeOf(this, BadTimeError.prototype);
  }

  get message(): string {
    const withValue = this.#value ? `"${this.#value}" is an ` : "";
    const withKey = this.#key ? `, key: ${this.#key}` : "";
    return `${withValue}invalid time${withKey}`;
  }
  set key(key: string) {
    this.#key = key;
  }
  get key(): string {
    return this.#key;
  }
  get value(): string {
    return this.#value;
  }
}

export class BadDateError extends MyError {
  #key = "";
  #value = "";

  constructor(badDate?: string) {
    super();
    this.#value = badDate ?? "";
    Object.setPrototypeOf(this, BadDateError.prototype);
  }

  get message(): string {
    const withValue = this.#value ? `"${this.#value}" is an ` : "";
    const withKey = this.#key ? `, key: ${this.#key}` : "";
    return `${withValue}invalid date${withKey}`;
  }
  set key(key: string) {
    this.#key = key;
  }
  get key(): string {
    return this.#key;
  }
  get value(): string {
    return this.#value;
  }
}

export class BadDurationError extends MyError {
  #key = "";
  #value = "";

  constructor(badDuration?: string) {
    super();
    this.#value = badDuration ?? "";
    Object.setPrototypeOf(this, BadDurationError.prototype);
  }

  get message(): string {
    const withValue = this.#value ? `"${this.#value}" is an ` : "";
    const withKey = this.#key ? `, key: ${this.#key}` : "";
    return `${withValue}invalid duration${withKey}. Give a valid duration string or skip duration with a "."`;
  }
  set key(key: string) {
    this.#key = key;
  }
  get key(): string {
    return this.#key;
  }
  get value(): string {
    return this.#value;
  }
}

export class BadTimezoneError extends MyError {
  constructor(m: unknown) {
    super(m);
    Object.setPrototypeOf(this, BadTimezoneError.prototype);
  }
}

export class BadStateError extends MyError {
  constructor(m: unknown) {
    super(m);
    Object.setPrototypeOf(this, BadStateError.prototype);
  }
}

export class MigrationError extends MyError {
  constructor(m: unknown) {
    super(m);
    Object.setPrototypeOf(this, MigrationError.prototype);
  }
}

export class MergeError extends MyError {
  constructor(m: unknown) {
    super(m);
    Object.setPrototypeOf(this, MergeError.prototype);
  }
}

export class NotImplementedError extends MyError {
  constructor(m: unknown) {
    super(m);
    Object.setPrototypeOf(this, NotImplementedError.prototype);
  }
}

export type CouchDbError = {
  error: string;
  reason: string;
  status: number;
  name: string;
  message: string;
};

export const isCouchDbError = (error: unknown): error is CouchDbError => {
  return !!(error as CouchDbError).error && !!(error as CouchDbError).reason;
};

export class DatumViewMissingError extends MyError {
  constructor(map_name?: unknown) {
    super(
      `Missing internal datum view ${map_name}. Please run setup on this database`,
    );
    Object.setPrototypeOf(this, DatumViewMissingError.prototype);
  }
}

export class LastDocsTooOldError extends MyError {
  constructor(m: unknown) {
    super(m);
    Object.setPrototypeOf(this, LastDocsTooOldError.prototype);
  }
}

export class FieldError extends MyError {
  constructor(m: unknown) {
    super(m);
    Object.setPrototypeOf(this, FieldError.prototype);
  }
}
