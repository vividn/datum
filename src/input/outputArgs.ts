import yargs, { Argv } from "yargs";

export type OutputArgs = {
  showAll?: boolean;
  show?: Show;
  formatString?: string;
};

export enum Show {
  Default = "default",
  None = "none",
  Minimal = "minimal",
  Standard = "standard",
  Format = "format",
  All = "all",
}

export function outputYargs(otherYargs?: Argv): Argv {
  const yarg = otherYargs ?? yargs;
  return yarg.group(["show", "showAll", "formatString"], "Output").options({
    "show-all": {
      describe: "Show complete document when displaying, not just data",
      type: "boolean",
      alias: "A",
    },
    show: {
      describe: "how much of documents to show",
      type: "string",
      choices: Object.values(Show),
      default: "default",
      conflict: "show-all",
    },
    "format-string": {
      describe:
        "create a custom output string for visualizing the doc(s). Specify %keys% with percent signs",
      type: "string",
      alias: "o",
    },
  });
}
