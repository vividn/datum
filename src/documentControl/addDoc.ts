import {
  EitherDocument,
  EitherPayload,
  isDatumDocument,
  isDatumPayload,
} from "./DatumDocument";
import { DateTime } from "luxon";
import { IdError, isCouchDbError } from "../errors";
import { jClone } from "../utils/jClone";
import { UpdateStrategyNames } from "./combineData";
import { updateDoc } from "./updateDoc";
import { showCreate, showExists, showFailed } from "../output/output";
import { BaseDocControlArgs, DocExistsError } from "./base";
import isEqual from "lodash.isequal";
import { overwriteDoc } from "./overwriteDoc";
import { deleteDoc } from "./deleteDoc";
import {
  DataOrDesignDocument,
  isViewDocument,
  isViewPayload,
  ViewPayload,
} from "../views/DatumView";
import { assembleId } from "../ids/assembleId";
import { toDatumTime } from "../time/datumTime";

function payloadMatchesDbData(
  payload: EitherPayload,
  existingDoc: EitherDocument,
) {
  const existingWithoutRev = jClone(existingDoc) as EitherPayload;
  delete existingWithoutRev._rev;

  return (
    (isDatumPayload(payload) &&
      isDatumDocument(existingDoc) &&
      isEqual(payload.data, existingDoc.data)) ||
    (isViewPayload(payload) &&
      isViewDocument(existingDoc) &&
      isEqual(payload.views, existingDoc.views)) ||
    (!isDatumPayload(payload) &&
      !isDatumDocument(existingDoc) &&
      isEqual(payload, existingWithoutRev))
  );
}

export type ConflictStrategyNames =
  | UpdateStrategyNames
  | "overwrite"
  | "delete"
  | "fail";

const conflictRecord: Record<ConflictStrategyNames, any> = {
  merge: "",
  useOld: "",
  preferOld: "",
  update: "",
  preferNew: "",
  useNew: "",
  removeConflicting: "",
  xor: "",
  intersection: "",
  append: "",
  prepend: "",
  appendSort: "",
  mergeSort: "",
  overwrite: "",
  delete: "",
  fail: "",
};
export const conflictChoices = Object.keys(conflictRecord);

type addDocType = {
  payload: EitherPayload | ViewPayload;
  conflictStrategy?: ConflictStrategyNames;
} & BaseDocControlArgs;

export async function addDoc({
  db,
  payload,
  conflictStrategy,
  outputArgs = {},
}: addDocType): Promise<DataOrDesignDocument> {
  payload = jClone(payload);
  let id;
  if (payload.meta) {
    const now = toDatumTime(DateTime.local());
    payload.meta.createTime = now;
    payload.meta.modifyTime = now;

    // reassemble id in case it depends upon createTime or modifyTime
    id = assembleId({ payload });

    payload._id = id;

    // Don't care about _rev for adding, will detect and redirect conflicts manually
    delete payload._rev;
  } else {
    id = payload._id;
    if (id === undefined) {
      throw new IdError("id could not be determined");
    }
  }
  try {
    await db.put(payload);
  } catch (error) {
    if (isCouchDbError(error) && error.name !== "conflict") {
      throw error;
    }
    const existingDoc = await db.get(id);

    if (conflictStrategy !== undefined && conflictStrategy !== "fail") {
      if (conflictStrategy === "overwrite") {
        return overwriteDoc({ db, id, payload, outputArgs: outputArgs });
      }
      if (conflictStrategy === "delete") {
        return deleteDoc({ db, id, outputArgs: outputArgs });
      }
      return await updateDoc({
        db,
        id,
        payload,
        updateStrategy: conflictStrategy,
        outputArgs: outputArgs,
      });
    }
    if (payloadMatchesDbData(payload, existingDoc)) {
      showExists(existingDoc, outputArgs);
      return existingDoc;
    } else {
      showFailed(payload, outputArgs);
      showExists(existingDoc, outputArgs);
      throw new DocExistsError(payload, existingDoc);
    }
  }
  const addedDoc = await db.get(id);
  showCreate(addedDoc, outputArgs);
  return addedDoc;
}
