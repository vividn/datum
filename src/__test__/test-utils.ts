import { CouchDbError } from "../errors";
import {
  DataOnlyDocument,
  DatumData,
  DatumDocument,
  DatumMetadata,
  EitherDocument,
  EitherPayload,
} from "../documentControl/DatumDocument";
import Mock = jest.Mock;
import { DateTime, Settings } from "luxon";
import { parseTimeStr } from "../time/parseTimeStr";
import { now } from "../time/timeUtils";
import { connectDb } from "../auth/connectDb";
import * as connectDbModule from "../auth/connectDb";
import { defaultIdComponents } from "../ids/defaultIdComponents";
import { buildIdStructure } from "../ids/buildIdStructure";
import { defaults } from "../input/defaults";
import { assembleId } from "../ids/assembleId";
import * as newHumanIdModule from "../meta/newHumanId";

// eslint-disable-next-line @typescript-eslint/no-empty-function
export const pass = (): void => {};
export const fail = (): never => {
  throw Error;
};

export const mockDocMissingError: CouchDbError = {
  status: 404,
  name: "not_found",
  error: "not_found",
  reason: "missing",
  message: "missing",
};

export const mockDocDeletedError: CouchDbError = {
  status: 404,
  message: "deleted",
  name: "not_found",
  error: "not_found",
  reason: "deleted",
};

export const mockMissingNamedViewError: CouchDbError = {
  status: 404,
  name: "not_found",
  error: "not_found",
  reason: "missing_named_view",
  message: "missing_named_view",
};

export async function resetTestDb(
  db: PouchDB.Database & {
    __opts?: PouchDB.Configuration.DatabaseConfiguration;
    _destroyed?: boolean;
    _closed?: boolean;
  },
): Promise<PouchDB.Database> {
  await db.destroy().catch(pass);
  // nasty hack to reopen closed database
  delete db._destroyed;
  delete db._closed;
  return db.constructor(db.name, db.__opts);
}

export function testDbLifecycle(
  dbName: string,
): PouchDB.Database<EitherPayload> {
  const db = connectDb({ db: dbName });

  beforeEach(async () => {
    await resetTestDb(db);
    jest.spyOn(connectDbModule, "connectDb").mockReturnValue(db);
  });

  afterEach(async () => {
    await db.destroy().catch(pass);
  });

  return db;
}

export function mockedLogLifecycle(): Mock {
  const originalLog = console.log;
  const mockedLog = jest.fn() as Mock;

  beforeEach(async () => {
    console.log = mockedLog;
  });

  afterEach(async () => {
    console.log = originalLog;
    mockedLog.mockReset();
  });

  return mockedLog;
}

const originalNowFn = Settings.now;
const nowStack: (() => number)[] = [];
export function setNow(timeStr: string): DateTime {
  const parsedTime = parseTimeStr({ timeStr, referenceTime: DateTime.now() });
  const newNow = () => parsedTime.toMillis();
  Settings.now = newNow;
  return parsedTime;
}

export function pushNow(timeStr: string): DateTime {
  nowStack.push(Settings.now);
  return setNow(timeStr);
}

export function popNow(): DateTime {
  const lastNow = nowStack.pop();
  if (lastNow === undefined) {
    throw new Error("tried to pop non existent now");
  }
  Settings.now = lastNow;
  return now();
}

export function restoreNow(): DateTime {
  Settings.now = originalNowFn;
  nowStack.length = 0;
  return now();
}

export function at<A extends any[], O>(
  timeStr: string,
  fn: (...args: A) => Promise<O>,
): (...args: A) => Promise<O> {
  return async (...args: A): Promise<O> => {
    pushNow(timeStr);
    const returnVal = await fn(...args);
    popNow();
    return returnVal;
  };
}

// TODO: Transition all tests to use makeDoc where appropriate
export function makeDoc(
  data: DatumData,
  meta: DatumMetadata | false = {},
  include_rev = false,
): EitherDocument {
  let doc: EitherDocument;
  if (meta === false) {
    doc = { ...data } as DataOnlyDocument;
  } else {
    doc = { ...data, meta } as DatumDocument;
  }

  meta = meta || {};
  if (doc._id === undefined) {
    let idStructure: string;
    if (meta?.idStructure) {
      idStructure = meta.idStructure;
    } else {
      const { defaultIdParts, defaultPartitionParts } = defaultIdComponents({
        data,
      });
      idStructure = buildIdStructure({
        idParts: defaultIdParts,
        delimiter: defaults.idDelimiter,
        partition: defaultPartitionParts,
      });
    }
    doc._id = assembleId({ payload: doc, idStructure });
  }

  if (include_rev) {
    doc._rev = "1-foo";
  }
  return doc;
}

export async function deterministicHumanIds(seed?: number): Promise<void> {
  let a = seed || 20231018;
  function random() {
    const x = Math.sin(a++) * 10000;
    return x - Math.floor(x);
  }
  function mockNewHumanId(): string {
    return random().toString(36).slice(2) + random().toString(36).slice(2);
  }

  beforeEach(() => {
    a = seed || 20231018;
    jest
      .spyOn(newHumanIdModule, "newHumanId")
      .mockImplementation(mockNewHumanId);
  });
}

// export async function generateSampleDay(dateStr = "2022-08-14") {
//
//   setNow(dateStr);
//   at('8:30', addCmd)({field: "sleep", })
//
//   popNow();
// }
