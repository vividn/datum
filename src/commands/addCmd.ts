import { EitherDocument } from "../documentControl/DatumDocument";
import { connectDb } from "../auth/connectDb";
import {
  addDoc,
  conflictChoices,
  ConflictStrategyNames,
} from "../documentControl/addDoc";
import { dataArgs, DataArgs, handleDataArgs } from "../input/dataArgs";
import { MainDatumArgs } from "../input/mainArgs";
import { addIdAndMetadata } from "../meta/addIdAndMetadata";
import { primitiveUndo } from "../undo/primitiveUndo";
import { FieldArgs, fieldArgs } from "../input/fieldArgs";
import { flexiblePositional } from "../input/flexiblePositional";
import { updateLastDocsRef } from "../documentControl/lastDocs";
import { compileState } from "../state/compileState";
import { ArgumentParser } from "argparse";
import { dbArgs } from "../input/dbArgs";
import { outputArgs } from "../input/outputArgs";
import { parseIfNeeded } from "../utils/parseIfNeeded";

export const newDocArgs = new ArgumentParser({
  add_help: false,
});
newDocArgs.add_argument("--no-metadata", "-M", {
  help: "do not include meta data in document",
  action: "store_true",
  dest: "noMetadata",
});
newDocArgs.add_argument("--id-part", "--id", {
  help:
    "Which field(s) to use for the _id field in the document." +
    " Can either be a single string with fields delimited by --id-delimiter" +
    " or can be used multiple times to progressively assemble an id delimited by --id-delimiter",
  action: "append",
  dest: "idParts",
});
newDocArgs.add_argument("--id-delimiter", {
  help: "spacer between fields in the id",
});
newDocArgs.add_argument("--undo", "-u", {
  help: "undoes the last datum entry",
  action: "store_true",
});
newDocArgs.add_argument("--force-undo", "-U", {
  help: "forces an undo, even if the datapoint was entered more than 15 minutes ago",
  action: "store_true",
  dest: "forceUndo",
});
newDocArgs.add_argument("--merge", "-x", {
  help: "on conflict with an existing document update with the merge strategy. Equivalent to `--update merge`",
  action: "store_const",
  const: "merge",
  dest: "conflict",
});
newDocArgs.add_argument("--conflict", "-X", {
  help: `on conflict, update with given strategy.`,
  choices: conflictChoices,
});

export const addArgs = new ArgumentParser({
  add_help: false,
  parents: [fieldArgs, newDocArgs, dataArgs],
});
export const addCmdArgs = new ArgumentParser({
  description: "add a document",
  prog: "datum add",
  usage: `%(prog)s <field> [data..]
  %(prog)s --fieldless [data..]
  %(prog)s <field> -k <reqKey> -k <optKey>=[defaultValue] ... <reqValue> [optValue] [data..]
`,
  parents: [addArgs, outputArgs, dbArgs],
});

export type AddCmdArgs = MainDatumArgs &
  FieldArgs &
  DataArgs & {
    noMetadata?: boolean;
    idParts?: string[];
    idDelimiter?: string;
    undo?: boolean;
    forceUndo?: boolean;
    merge?: boolean;
    conflict?: ConflictStrategyNames;
  };

export async function addCmd(
  args: AddCmdArgs | string | string[],
  preparsed?: Partial<AddCmdArgs>,
): Promise<EitherDocument> {
  args = parseIfNeeded(addCmdArgs, args, preparsed);
  const db = connectDb(args);

  flexiblePositional(args, "field", "field", args.fieldless);

  const payloadData = handleDataArgs(args);

  const payloadWithState = await compileState(db, payloadData);

  const payload = addIdAndMetadata(payloadWithState, args);

  // update now in case the addDoc fails due to conflict
  await updateLastDocsRef(db, payload._id);

  const { undo, forceUndo } = args;
  if (undo || forceUndo) {
    return await primitiveUndo({
      db,
      payload,
      force: forceUndo,
      outputArgs: args,
    });
  }

  const conflictStrategy = args.conflict ?? (args.merge ? "merge" : undefined);
  const doc = await addDoc({
    db,
    payload,
    conflictStrategy,
    outputArgs: args,
  });

  // if addDoc has changed the id (e.g. because it relies on the modifiyTime), update lastDocRef again
  // TODO: if changing lastDoc to history may need to change this to overwrite first update
  if (doc._id !== payload._id) {
    await updateLastDocsRef(db, doc._id);
  }

  return doc;
}
