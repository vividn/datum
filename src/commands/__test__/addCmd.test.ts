import { fail, mockedLogLifecycle, testDbLifecycle } from "../../test-utils";
import { BaseDataError } from "../../errors";
import { DatumDocument } from "../../documentControl/DatumDocument";
import { addCmd } from "../addCmd";
import * as addDoc from "../../documentControl/addDoc";
import { DocExistsError } from "../../documentControl/base";
import SpyInstance = jest.SpyInstance;
import { Show } from "../../input/outputArgs";

describe("addCmd", () => {
  const dbName = "add_cmd_test";

  const db = testDbLifecycle(dbName);

  const mockedLog = mockedLogLifecycle();

  let addDocSpy: SpyInstance;
  beforeEach(() => {
    addDocSpy = jest.spyOn(addDoc, "addDoc");
  });

  it("inserts documents into couchdb", async () => {
    await addCmd({});

    await db.info().then((info) => {
      expect(info.doc_count).toEqual(1);
    });
  });

  it("calls addDoc", async () => {
    await addCmd({ idPart: "%foo%", data: ["foo=abc"] });
    const spyCall = addDocSpy.mock.calls[0][0];
    expect(spyCall).toMatchObject({
      db: db,
      payload: { data: { foo: "abc" }, meta: { idStructure: "%foo%" } },
    });
  });

  it("Can remove metadata entirely", async () => {
    expect(await addCmd({ idPart: "hasMetadata" })).toHaveProperty("meta");
    expect(
      await addCmd({ idPart: "noMeta", noMetadata: true })
    ).not.toHaveProperty("meta");
  });

  it("tells the user if the document already exists with identical data", async () => {
    await addCmd({
      idPart: "my name is bob",
      noTimestamp: true,
      data: ["foo=bar"],
      show: Show.Standard,
    });
    expect(mockedLog).toHaveBeenCalledWith(expect.stringContaining("CREATE"));
    expect(mockedLog).not.toHaveBeenCalledWith(
      expect.stringContaining("EXISTS")
    );

    mockedLog.mockReset();

    await addCmd({
      idPart: "my name is bob",
      noTimestamp: true,
      data: ["foo=bar"],
      show: Show.Standard,
    });
    expect(mockedLog).not.toHaveBeenCalledWith(
      expect.stringContaining("CREATE")
    );
    expect(mockedLog).toHaveBeenCalledWith(expect.stringContaining("EXISTS"));
  });

  it("fails if addedDocument conflicts with different data", async () => {
    await addCmd({
      idPart: "my name is doug",
      noTimestamp: true,
      data: ["foo=bar"],
      show: Show.Standard,
    });
    expect(mockedLog).toHaveBeenCalledWith(expect.stringContaining("CREATE"));
    expect(mockedLog).not.toHaveBeenCalledWith(
      expect.stringContaining("EXISTS")
    );

    mockedLog.mockReset();

    try {
      await addCmd({
        idPart: "my name is doug",
        noTimestamp: true,
        data: ["different=data"],
        show: Show.Standard,
      });
      fail();
    } catch (e) {
      expect(e).toBeInstanceOf(DocExistsError);
    }
    expect(mockedLog).toHaveBeenCalledWith(expect.stringContaining("FAILED"));
    expect(mockedLog).toHaveBeenCalledWith(expect.stringContaining("EXISTS"));
  });

  it("inserts id structure into the metadata", async () => {
    expect(
      await addCmd({ idPart: ["rawString", "%foo%!!"], data: ["foo=abc"] })
    ).toMatchObject({
      meta: { idStructure: "rawString__%foo%!!" },
    });
  });

  it("can use custom base data", async () => {
    expect(
      await addCmd({ baseData: "{a: 1, b:2, c:3 }", idPart: "basedata-doc1" })
    ).toMatchObject({ data: { a: 1, b: 2, c: 3 }, _id: "basedata-doc1" });
  });

  it("can write payloads directly by specifying base-data and no-metadata", async () => {
    expect(
      await addCmd({
        noMetadata: true,
        baseData: "{a: 1, b:2, c:3}",
        idPart: "basedata-doc2",
      })
    ).toMatchObject({
      _id: "basedata-doc2",
      a: 1,
      b: 2,
      c: 3,
    });
  });

  it("throws a BaseDataError if baseData is malformed", async () => {
    await expect(addCmd({ baseData: "string" })).rejects.toThrowError(
      BaseDataError
    );
  });

  it("prefers the _id specified when in no-metadata mode", async () => {
    expect(
      await addCmd({
        noMetadata: true,
        baseData: "{ _id: payload-id }",
        idPart: "argument-id",
      })
    ).toMatchObject({ _id: "payload-id" });
    expect(
      await addCmd({
        noMetadata: true,
        baseData: "{ _id: payload-id-2 }",
        idPart: "%keyId%",
        data: ["keyId=key-id"],
      })
    ).toMatchObject({ _id: "payload-id-2" });
    expect(
      await addCmd({
        noMetadata: true,
        data: ["_id=posArgs-id"],
        idPart: "idPart-id",
      })
    ).toMatchObject({ _id: "posArgs-id" });
  });

  it("does not contain idStructure in the metadata if id does not depend on values from data", async () => {
    const returnDoc = (await addCmd({ idPart: "notAField" })) as DatumDocument;
    expect(returnDoc._id).toBe("notAField");
    expect(returnDoc.meta).not.toHaveProperty("idStructure");
  });

  it("contains random identifiers in the metadata", async () => {
    const doc = await addCmd({});
    const { random, humanId } = doc?.meta;

    expect(random).toBeGreaterThanOrEqual(0);
    expect(random).toBeLessThanOrEqual(1);
    expect(humanId).toEqual(expect.stringMatching(/^[0-9a-z]+$/));
  });

  it("can display just the data of documents or the whole documents", async () => {
    const matchExtraKeysInAnyOrder =
      /^(?=[\s\S]*_id:)(?=[\s\S]*data:)(?=[\s\S]*meta:)/;
    await addCmd({ idPart: "this-id" });
    expect(mockedLog).not.toHaveBeenCalledWith(
      expect.stringMatching(matchExtraKeysInAnyOrder)
    );

    mockedLog.mockClear();

    await addCmd({ idPart: "that-id", showAll: true });
    expect(mockedLog).toHaveBeenCalledWith(
      expect.stringMatching(matchExtraKeysInAnyOrder)
    );
  });

  it("can merge into an existing document with --merge", async () => {
    await addCmd({ idPart: "doc-id", data: ["foo=abc"] });
    const newDoc = await addCmd({
      idPart: "doc-id",
      data: ["foo=def"],
      merge: true,
    });
    expect(newDoc).toMatchObject({ data: { foo: ["abc", "def"] } });
    expect(addDocSpy).toHaveBeenCalledTimes(2);
    expect(addDocSpy.mock.calls[1][0].conflictStrategy).toEqual("merge");
  });

  it("can update and existing document with --conflict", async () => {
    await addCmd({ idPart: "doc-id", data: ["foo=abc"] });
    const newDoc = await addCmd({
      idPart: "doc-id",
      data: ["foo=def"],
      conflict: "preferNew",
    });
    expect(addDocSpy).toHaveBeenCalledTimes(2);
    expect(addDocSpy.mock.calls[1][0].conflictStrategy).toEqual("preferNew");
    expect(newDoc).toMatchObject({ data: { foo: "def" } });
  });

  // TODO: write tests for all of the various options

  it("stores the occurTime in the data", async () => {
    const newDoc = (await addCmd({
      date: "2021-08-23",
      time: "12",
      timezone: "0",
    })) as DatumDocument;
    expect(newDoc.data).toHaveProperty("occurTime", "2021-08-23T12:00:00.000Z");
    expect(newDoc.meta).not.toHaveProperty("occurTime");
  });

  it("stores the occurTime and utcOffset in DataOnly docs", async () => {
    const newDoc = (await addCmd({
      noMetadata: true,
      date: "2021-08-23",
      time: "12",
      timezone: "0",
    })) as DatumDocument;
    expect(newDoc).toHaveProperty("occurTime", "2021-08-23T12:00:00.000Z");
    expect(newDoc).toHaveProperty("occurUtcOffset", 0);
  });

  it("stores utcOffset", async () => {
    const newDoc = (await addCmd({
      date: "2021-08-23",
      time: "12",
      timezone: "0",
    })) as DatumDocument;
    expect(newDoc.data).toHaveProperty("occurUtcOffset", 0);
  });
});
