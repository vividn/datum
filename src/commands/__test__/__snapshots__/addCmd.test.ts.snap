// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addCmd change command can become a start command by having start as a trailing word 1`] = `
{
  "_id": "field:2023-12-21T14:00:00.000Z",
  "_rev": Any<String>,
  "data": {
    "dur": "PT30M",
    "field": "field",
    "lastState": null,
    "occurTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
    "opt1": "optVal",
    "req1": "reqVal",
    "state": true,
  },
  "meta": {
    "createTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
    "humanId": "elfuv1633zeysuyhc0l9ss",
    "idStructure": "%field%:%occurTime%",
    "modifyTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
  },
}
`;

exports[`addCmd change command can become a switch command by having start as a trailing word 1`] = `
{
  "_id": "field:2023-12-21T14:00:00.000Z",
  "_rev": Any<String>,
  "data": {
    "dur": "PT5M30S",
    "field": "field",
    "lastState": null,
    "occurTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
    "opt1": "optVal",
    "req1": "reqVal",
    "state": "stateName",
  },
  "meta": {
    "createTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
    "humanId": "elfuv1633zeysuyhc0l9ss",
    "idStructure": "%field%:%occurTime%",
    "modifyTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
  },
}
`;

exports[`addCmd change command can become an end command by having start as a trailing word 1`] = `
{
  "_id": "field:2023-12-21T14:00:00.000Z",
  "_rev": Any<String>,
  "data": {
    "dur": "PT30M",
    "field": "field",
    "lastState": null,
    "occurTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
    "opt1": "optVal",
    "req1": "reqVal",
    "state": false,
  },
  "meta": {
    "createTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
    "humanId": "elfuv1633zeysuyhc0l9ss",
    "idStructure": "%field%:%occurTime%",
    "modifyTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
  },
}
`;

exports[`addCmd change command can become an occur command by having start as a trailing word 1`] = `
{
  "_id": "field:2023-12-21T14:00:00.000Z",
  "_rev": Any<String>,
  "data": {
    "dur": null,
    "field": "field",
    "lastState": null,
    "occurTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
    "opt1": "optVal",
    "req1": "reqVal",
  },
  "meta": {
    "createTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
    "humanId": "elfuv1633zeysuyhc0l9ss",
    "idStructure": "%field%:%occurTime%",
    "modifyTime": {
      "o": 0,
      "tz": "UTC",
      "utc": "2023-12-21T14:00:00.000Z",
    },
  },
}
`;
